File: vlingo-hike-backend/src/test/java/io/vlingo/hike/InMemoryJournalListener.java
package io.vlingo.hike;

import io.vlingo.actors.Actor;
import io.vlingo.actors.testkit.TestUntil;
import io.vlingo.symbio.Entry;
import io.vlingo.symbio.State;
import io.vlingo.symbio.store.journal.JournalListener;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

public class InMemoryJournalListener implements JournalListener<String> {
    public final List<Entry<String>> allEntries = new CopyOnWriteArrayList<>();
    public final List<State<String>> allSnapshots = new CopyOnWriteArrayList<>();
    public TestUntil until = TestUntil.happenings(1);

    @Override
    public void appended(final Entry<String> entry) {
        allEntries.add(entry);
        allSnapshots.add(null);
        until.happened();
    }

    @Override
    public void appendedWith(final Entry<String> entry, final State<String> snapshot) {
        allEntries.add(entry);
        allSnapshots.add(snapshot);
        until.happened();
    }

    @Override
    public void appendedAll(final List<Entry<String>> entries) {
        allEntries.addAll(entries);
        final List<State<String>> snapshots = new ArrayList<>(entries.size());
        for (int idx = 0; idx < entries.size(); ++idx) {
            snapshots.add(null);
        }
        allSnapshots.addAll(snapshots);
        until.happened();
    }

    @Override
    public void appendedAllWith(final List<Entry<String>> entries, final State<String> snapshot) {
        allEntries.addAll(entries);
        final List<State<String>> snapshots = new ArrayList<>(entries.size());
        for (int idx = 0; idx < (entries.size() - 1); ++idx) {
            snapshots.add(null);
        }
        snapshots.add(snapshot);
        allSnapshots.addAll(snapshots);
        until.happened();
    }

    public int confirmExpectedEntries(final int count, final int retries) {
        for (int idx = 0; idx < retries; ++idx) {
            if (allEntries.size() == count) {
                return count;
            }
            try { Thread.sleep(100); } catch (Exception e) { }
        }
        return allEntries.size();
    }
}

File: xoom-hike-backend/src/test/kotlin/io/vlingo/hike/InMemoryJournalListener.kt
package io.vlingo.hike

import io.vlingo.actors.testkit.TestUntil
import io.vlingo.symbio.Entry
import io.vlingo.symbio.State
import io.vlingo.symbio.store.journal.JournalListener

import java.util.ArrayList
import java.util.concurrent.CopyOnWriteArrayList

class InMemoryJournalListener : JournalListener<String> {
    val allEntries: MutableList<Entry<String>?> = CopyOnWriteArrayList()
    val allSnapshots: MutableList<State<String>?> = CopyOnWriteArrayList()
    var until = TestUntil.happenings(1)

    override fun appended(entry: Entry<String>?) {
        allEntries.add(entry)
        allSnapshots.add(null)
        until.happened()
    }

    override fun appendedWith(entry: Entry<String>?, snapshot: State<String>?) {
        allEntries.add(entry)
        allSnapshots.add(snapshot)
        until.happened()
    }

    override fun appendedAll(entries: List<Entry<String>>?) {
        allEntries.addAll(entries ?: emptyList())
        val snapshots = ArrayList<State<String>?>(entries?.size ?: 0)
        for (idx in entries?.indices ?: emptyList<Int>()) {
            snapshots.add(null)
        }
        allSnapshots.addAll(snapshots)
        until.happened()
    }

    override fun appendedAllWith(entries: List<Entry<String>>?, snapshot: State<String>?) {
        allEntries.addAll(entries ?: emptyList())
        val snapshots = ArrayList<State<String>?>(entries?.size ?: 0)
        for (idx in 0 until ((entries?.size ?: 0) - 1)) {
            snapshots.add(null)
        }
        snapshots.add(snapshot)
        allSnapshots.addAll(snapshots)
        until.happened()
    }
}